services/kubernetesmisc/src/handlers/kubernetesBuildCancel.ts

```diff
@@ -0,0 +1,135 @@
+  }
+};
+
+const getUrlTokenFromProjectInfo = (projectOpenShift, name) => {
+  try {
+    const url = projectOpenShift.openshift.consoleUrl.replace(/\/$/, '');
+    const token = projectOpenShift.openshift.token || '';
+    return { url, token };
+  } catch (error) {
+    logger.warn(
+      `Error while loading information for project ${name}: ${error}`
+    );
+    throw error;
+  }
+};
+
+const getConfig = (url, token) => ({
+  url,
+  insecureSkipTlsVerify: true,
+  auth: {
+    bearer: token
+  }
+});
+
+const getJobInfo = async (client: Api.ApiRoot, namespace: string, jobName: string) => {
+  let result;
+  try {
+    result = await client.apis.batch.v1.namespaces(namespace).jobs(jobName).get()
+  } catch (err) {
+    if (err.code == 404) {
+      return undefined;
+    } else {
+      logger.error(err);
+      throw new Error();
+    }
+  }
+  return result;
+}
+
+const updateLagoonTask = async (jobInfo, jobStatus, taskId, project, jobName) => {
+  // Update lagoon task
+  try {
+    const convertDateFormat = R.init;
+    const dateOrNull = R.unless(R.isNil, convertDateFormat);
+    let completedDate = dateOrNull(jobInfo.body.status.completionTime);
+
+    if (jobStatus === 'failed') {
+      completedDate = dateOrNull(jobInfo.body.status.conditions[0].lastTransitionTime);
+    }
+
+    await updateTask(taskId, {
+      status: jobStatus.toUpperCase(),
+      created: convertDateFormat(jobInfo.body.metadata.creationTimestamp),
+      started: dateOrNull(jobInfo.body.status.startTime),
+      completed: completedDate
+    });
+  } catch (error) {
+    logger.error(
+      `Could not update task ${project.name} ${jobName}. Message: ${error}`
+    );
+  }
+}
+
+const deleteJob = async (client: Api.ApiRoot, namespace: string, jobName: string) => {
+  try {
+    const options = {
+      body: {
+        kind: 'DeleteOptions',
+        apiVersion: 'v1',
+        propagationPolicy: 'Foreground',
+      }
+    };
+    // https://github.com/godaddy/kubernetes-client/blob/master/docs/1.13/Job.md#apisbatchv1namespacesnamespacejobsnamedelete
+    return await client.apis.batch.v1.namespaces(namespace).jobs(jobName).delete(options)
+  } catch (err) {
+    logger.error(`Couldn't delete job ${jobName}. Error: ${err}`);
+  }
+};
+
+const kubernetesBuildCancel = async (data: any) => {
+  const { build: { name: buildName, id }, project, environment } = data;
+
+  const { project: projectInfo } = await getOpenShiftInfoForProject(project.name);
+  const { url, token } = getUrlTokenFromProjectInfo(projectInfo, project.name);
+  const config: ClientConfiguration = getConfig(url, token);
+  const client = new Client({ config });
+  const { namespace } = generateSanitizedNames(project, environment, projectInfo);
+
+  // Check that job is still active
+  const jobInfo = await getJobInfo(client, namespace, buildName);
+  if (jobInfo) {
+    logger.error(`Job ${buildName} does not exist, bailing`);
+    await deleteJob(client, namespace, buildName);
+  }
```

Comment:
I initially thought to check if the status is "active," but since the jobs monitor will (or is supposed to) delete jobs after they're failed/complete, then I guess just checking that the job exists at all is the same as checking if it's active? Can you add a comment here to explain that? It's def a business logic condition that wouldn't be clear unless you were intimately familiar with the whole tasks/jobs system.

Reply:
👍